cpp/include/raft/spectral/sm_utils.hpp

```diff
@@ -0,0 +1,435 @@
+  return *reinterpret_cast<long long *>(&a);
+#endif
+#else
+  return 0.0;
+#endif
+}
+
+// specifically for triangles counting
+static __device__ __forceinline__ uint64_t shfl_down(uint64_t r, int offset,
+                                                     int bound = 32,
+                                                     int mask = DEFAULT_MASK) {
+#if __CUDA_ARCH__ >= 300
+#ifdef USE_CG
+  int2 a = *reinterpret_cast<int2 *>(&r);
+  a.x = __shfl_down_sync(mask, a.x, offset, bound);
+  a.y = __shfl_down_sync(mask, a.y, offset, bound);
+  return *reinterpret_cast<uint64_t *>(&a);
+#else
+  int2 a = *reinterpret_cast<int2 *>(&r);
+  a.x = __shfl_down(mask, a.x, offset, bound);
+  a.y = __shfl_down(mask, a.y, offset, bound);
+  return *reinterpret_cast<uint64_t *>(&a);
+#endif
+#else
+  return 0.0;
+#endif
+}
+
+static __device__ __forceinline__ int shfl_up(int r, int offset, int bound = 32,
+                                              int mask = DEFAULT_MASK) {
+#if __CUDA_ARCH__ >= 300
+#ifdef USE_CG
+  return __shfl_up_sync(mask, r, offset, bound);
+#else
+  return __shfl_up(r, offset, bound);
+#endif
+#else
+  return 0.0f;
+#endif
+}
+
+static __device__ __forceinline__ float shfl_up(float r, int offset,
+                                                int bound = 32,
+                                                int mask = DEFAULT_MASK) {
+#if __CUDA_ARCH__ >= 300
+#ifdef USE_CG
+  return __shfl_up_sync(mask, r, offset, bound);
+#else
+  return __shfl_up(r, offset, bound);
+#endif
+#else
+  return 0.0f;
+#endif
+}
+
+static __device__ __forceinline__ double shfl_up(double r, int offset,
+                                                 int bound = 32,
+                                                 int mask = DEFAULT_MASK) {
+#if __CUDA_ARCH__ >= 300
+#ifdef USE_CG
+  int2 a = *reinterpret_cast<int2 *>(&r);
+  a.x = __shfl_up_sync(mask, a.x, offset, bound);
+  a.y = __shfl_up_sync(mask, a.y, offset, bound);
+  return *reinterpret_cast<double *>(&a);
+#else
+  int2 a = *reinterpret_cast<int2 *>(&r);
+  a.x = __shfl_up(a.x, offset, bound);
+  a.y = __shfl_up(a.y, offset, bound);
+  return *reinterpret_cast<double *>(&a);
+#endif
+#else
+  return 0.0;
+#endif
+}
+
+static __device__ __forceinline__ long long shfl_up(long long r, int offset,
+                                                    int bound = 32,
+                                                    int mask = DEFAULT_MASK) {
+#if __CUDA_ARCH__ >= 300
+#ifdef USE_CG
+  int2 a = *reinterpret_cast<int2 *>(&r);
+  a.x = __shfl_up_sync(mask, a.x, offset, bound);
+  a.y = __shfl_up_sync(mask, a.y, offset, bound);
+  return *reinterpret_cast<long long *>(&a);
+#else
+  int2 a = *reinterpret_cast<int2 *>(&r);
+  a.x = __shfl_up(a.x, offset, bound);
+  a.y = __shfl_up(a.y, offset, bound);
+  return *reinterpret_cast<long long *>(&a);
+#endif
+#else
+  return 0.0;
+#endif
+}
+
+static __inline__ __device__ double atomicFPAdd(double *addr, double val) {
```

Comment:
I don't know enough about raft's general design directions. I placed this header in `spectral/` because I assumed limited use (spectral and spectral-related functionality in `cugraph`). It doesn't seem to me there's enough reason to reuse these outside of spectral clustering. In fact, I suspect many would be deprecated / replaced in the (near) future.

But, if desired, I can move it somewhere. `raft/utils` (directory?) doesn't currently exist. Please specify exactly where this should be moved, because when that happens it will trigger necessary updates in the two other spectral clustering-related PRs. So, if necessary, I'd rather do this only once, rather than "experiment" with different locations, etc.

Reply:
IOW, if `raft/utils/sm_utils.hpp` _is_ the desired location then I will create the `utils` directory and move it. But, I'd prefer reviewers to be _sure_ that that is desired course of action, here. Thanks in advance.

Reply:
I think `raft/utils/sm_utils.hpp` is a great location. I've had to define some of these in the past and it's nice to have them around when needed (also so there isn't duplication when they are needed).

Reply:
Done.